contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,462 @@
+    fn _upper_binary_lookup(
+        &self,
+        key: U96,
+        mut low: U256,
+        mut high: U256,
+    ) -> U256 {
+        while low < high {
+            let mid = low.average(high);
+            if self._access(mid)._key.get() > key {
+                high = mid;
+            } else {
+                low = mid + uint!(1_U256);
+            }
+        }
+        high
+    }
+
+    /// Return the index of the first (oldest) checkpoint with key is
+    /// greater or equal than the search key, or `high` if there is none.
+    /// Indexes `low` and `high` define a section where to do the search, with
+    /// inclusive `low` and exclusive `high`.
+    ///
+    /// WARNING: `high` should not be greater than the array's length.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the checkpoint's state.
+    /// * `key` - Checkpoint key to lookup.
+    /// * `low` - inclusive index where search begins.
+    /// * `high` - exclusive index where search ends.
+    fn _lower_binary_lookup(
+        &self,
+        key: U96,
+        mut low: U256,
+        mut high: U256,
+    ) -> U256 {
+        while low < high {
+            let mid = low.average(high);
+            if self._access(mid)._key.get() < key {
+                low = mid + uint!(1_U256);
+            } else {
+                high = mid;
+            }
+        }
+        high
+    }
+
+    /// Immutable access on an element of the checkpoint's array.
+    /// The position is assumed to be within bounds.
+    /// Panic when out of bounds.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the checkpoint's state.
+    /// * `pos` - index of the checkpoint.
+    fn _access(&self, pos: U256) -> StorageGuard<Checkpoint160> {
+        self._checkpoints
+            .get(pos)
+            .unwrap_or_else(|| panic!("should get checkpoint at index `{pos}`"))
+    }
+
+    /// Mutable access on an element of the checkpoint's array.
+    /// The position is assumed to be within bounds.
+    /// Panic when out of bounds.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the checkpoint's state.
+    /// * `pos` - index of the checkpoint.
+    fn _access_mut(&mut self, pos: U256) -> StorageGuardMut<Checkpoint160> {
+        self._checkpoints
+            .setter(pos)
+            .unwrap_or_else(|| panic!("should get checkpoint at index `{pos}`"))
+    }
+
+    /// Append checkpoint without checking if sorted order pertains after.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the checkpoint's state.
+    /// * `key` - Checkpoint key to insert.
+    /// * `value` - Checkpoint value corresponding to insertion `key`.
+    fn _unchecked_push(&mut self, key: U96, value: U160) {
+        let mut new_checkpoint = self._checkpoints.grow();
+        new_checkpoint._key.set(key);
+        new_checkpoint._value.set(value);
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
```

Comment:
I am not sure if we have enough test coverage here.

Reply:
Yep. Highlighted it in my previous comment. Fill fix it